Add capability to display visual metrics

cliu55 opened the pull request:
Recent updates to browsertime lets us capture gecko profiler while
running the application. This change embeds visual metrics data into the
Gecko profile computed by browsertime, and lets the Firefox Profiler
display some of those metrics.

A visual metrics section is added to the profile meta dropdown, and a
new primary track is added to display graph of visual completeness  over
time, only if the information is available.

gregtatum commented:
@cliu55 Do you have any example profiles that you can include in this review? It'd be helpful to have a deploy preview showing what's going on here. For instance in #1883 you can see how Julien included different URLs to make it easy for the reviewer to see what's going on.

There is a deploy preview link available as well, if you click on `Details` next to `deploy/netlify`: https://deploy-preview-2054--perf-html.netlify.com/

Then you can manually put in an uploaded profile.

cliu55 commented:
Hi @gregtatum, below is a link to an example profile

https://deploy-preview-2054--perf-html.netlify.com/public/470874122ed3cebf33849a85b5145de1a2bd5421/calltree/?globalTrackOrder=5-4-0-1-2-3&hiddenGlobalTracks=1-2&localTrackOrderByPid=75565-1-2-0~75568-0~75566-0~75567-0-1~&thread=4&v=3

gregtatum commented on src/actions/profile-view.js:
```diff
@@ -250,6 +250,9 @@ export function selectTrack(trackReference: TrackReference): ThunkAction<void> {
         case 'screenshots':
```
These could be consolidated with

```
case 'screenshots':
case 'visualProgress':
  // Do not allow selecting these tracks.
  return;
```

cliu55 replied:
The cases are now consolidated.

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -99,20 +112,25 @@ class TrackMemoryCanvas extends React.PureComponent<CanvasProps> {
-      ctx.strokeStyle = ORANGE_50;
-      ctx.fillStyle = '#ff940088'; // Orange 50 with transparency.
+      ctx.strokeStyle = graphStrokeColor;
+      ctx.fillStyle = graphFillColor; // Blue 50 or Orange 50 with transparency.
```
I would remove this comment, since it will be easy to get out of date, and was only relevant with the hex value.

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -99,20 +112,25 @@ class TrackMemoryCanvas extends React.PureComponent<CanvasProps> {
+        // Y position for visual progress graph calculate based on sample count
+        // instead of accumulated count.
+        const unitGraphCount =
+          graphType === GRAPH_TYPE.visualProgress
```
Instead of doing an if/else check each iteration of this loop, I would add a `prop` value that can inject a function that will do this.

`computeGraphHeight: IndexIntoSamplesTable => DevicePixels`

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -121,12 +139,24 @@ class TrackMemoryCanvas extends React.PureComponent<CanvasProps> {
           ctx.moveTo(x, y);
         } else {
+          if (graphType === GRAPH_TYPE.visualProgress) {
+            // Plotting a extra point to generate a step-wise graph for visual progress.
```
```suggestion
            // Plot an extra point to generate a step-wise graph for visual progress.
```

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -25,12 +24,20 @@ import type {
 import type { ConnectedProps } from '../../utils/connect';
 
-import './TrackMemory.css';
+import './TrackCounter.css';
```
I'm torn on this patch, I initially recommended doing a copy/paste of this file, mainly because when a component has bimodal behavior, it gets a lot hard to get it right, and the component is much more complicated. I'd prefer a little code duplication over code complication.

If we keep this as a generic component, then I'd like to see the different behaviors encoded as "feature" checks, not "graph type" checks. I'm going to put my recommendations inline below.

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -44,14 +51,17 @@ type CanvasProps = {|
   +width: CssPixels,
   +height: CssPixels,
   +lineWidth: CssPixels,
+  +graphType: GraphType,
```
The `graphType` prop should be removed.

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -44,14 +51,17 @@ type CanvasProps = {|
   +height: CssPixels,
   +lineWidth: CssPixels,
+  +graphType: GraphType,
+  +graphStrokeColor: CanvasPattern,
```
Is this correct? Shouldn't is just be 'string'?

cliu55 replied:
My apologies, the type has been updated to string in the new changes

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -202,7 +240,7 @@ type State = {|
  * The memory track graph takes memory information from counters, and renders it as a
```
Please update this comment to reflect the general nature of the component, and remove references to the "memory track graph".

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -172,6 +205,11 @@ type OwnProps = {|
+  +graphStrokeColor: CanvasPattern,
+  +graphFillColor: string,
+  +graphDotColor: CanvasPattern,
+  +renderTooltip: any,
```
This needs a type, the `any` type means "opt out of type checking" which is dangerous. I believe the type is:

`(counterIndex: number) => React.Node`

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -121,12 +139,24 @@ class TrackMemoryCanvas extends React.PureComponent<CanvasProps> {
         }
+        yOld = y;
+      }
+      if (graphType === GRAPH_TYPE.visualProgress) {
```
I'm not quite sure what the best way to make this a generic feature rather than a graph type check. I think maybe we should just rely on the original behavior. Is there a reason not to? Also I think this implementation has an issue in that it will draw into the empty thread indicators.

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -121,12 +139,24 @@ class TrackMemoryCanvas extends React.PureComponent<CanvasProps> {
           // This is the first iteration, only move the line.
           ctx.moveTo(x, y);
         } else {
+          if (graphType === GRAPH_TYPE.visualProgress) {
```
I would recommend adding a prop:
`if (this.props.isStepwise)`

gregtatum commented on src/components/timeline/TrackCounterGraph.js:
```diff
@@ -282,21 +295,33 @@ class TrackMemoryGraphImpl extends React.PureComponent<Props, State> {
+    // For Visual Progress graph, use sample count instead of accumulated count.
     const unitSampleCount =
-      (accumulatedCounts[counterIndex] - minCount) / countRange;
+      graphType === GRAPH_TYPE.visualProgress
```
This should be computed through a function passed in as a prop.

gregtatum commented on src/components/timeline/TrackMemory.js:
```diff
@@ -75,10 +103,15 @@ export class TrackMemoryImpl extends React.PureComponent<Props, State> {
           graphHeight={TRACK_MEMORY_GRAPH_HEIGHT}
+          graphType={GRAPH_TYPE.memory}
+          graphStrokeColor={ORANGE_50}
+          graphFillColor="#ff940088"
```
The comment about the hex's color should be moved here.

gregtatum commented on src/components/timeline/TrackVisualProgress.js:
```diff
@@ -0,0 +1,112 @@
+
+type State = {||};
+
+export class TrackVisualProgressImpl extends React.PureComponent<Props, State> {
```
Please add some documentation to the top of this class explaining what's going on with it.

gregtatum replied:
I believe this comment still needs to be handled.

gregtatum commented on src/profile-logic/process-profile.js:
```diff
@@ -732,6 +733,79 @@ function _processCounters(
+    geckoProfile
+  );
+
+  visualProgressString.split(', ').forEach((visualProgress, index) => {
```
Please document what this `visualProgressString` content looks like.

cliu55 replied:
After some recent changes to browsertime and changes to my approach, visual progress data no longer needs to be processed and is ready to consume.

gregtatum commented on src/profile-logic/process-profile.js:
```diff
@@ -1014,6 +1088,10 @@ export function processProfile(
     );
   }
 
+  const visualProgress: Counter[] | null = geckoProfile.meta.visualMetrics
```
I think the visual progress is only using the counters data structure because it's convenient to use the memory track. I'm concerned that this approach needless complicates and confuses two concepts in order to not duplicate a bit of canvas code.

cliu55 replied:
I completely agree and I've made changes to my approach. Visual progress data now has it's own shape and a new component has been created to display the charts. Furthermore, visual progress data no longer needs to be processed and is ready to consume.

cliu55 commented on src/types/profile.js:
```diff
@@ -307,6 +307,19 @@ export type ExtensionTable = {|
 |};
 
+export type VisualMetrics = {|
+  SpeedIndex: number,
```
Hi Greg, it seems like the reason for the Pascal Casing here is because the data source is following suit with `visualmetrics.py` [upstream](https://github.com/sitespeedio/browsertime/blob/master/vendor/visualmetrics.py). Changing the visual metrics case style in `mozilla/browsertime` would make it difficult to upstream to `sitespeedio/browsertime`.

cliu55 commented:
Below is a link to a deploy preview

https://deploy-preview-2054--perf-html.netlify.com/public/7788671ed67417b9acd056310b99715892ffefe1/calltree/?globalTrackOrder=5-6-7-4-0-1-2-3&hiddenGlobalTracks=1-2&localTrackOrderByPid=21091-1-2-0~21094-0~21092-0~21093-0-1~&thread=4&v=4

cliu55 commented on src/types/profile.js:
```diff
@@ -307,6 +307,19 @@ export type ExtensionTable = {|
   length: number,
 |};
 
+export type VisualMetrics = {|
```
I completely agree. In the new changes, I've provided more context on what visual metrics are, where they are from, and URLs for more information.

cliu55 commented:
Hi @gregtatum, after reviewing your feedback on the pull request, I made some major changes and I am looking for another review. I replied to the relevant comments you left for the old changes, and I will also summarize some of my changes here.

1. You were concerned that visual progress data being made to look like counter in order to reuse charting code will cause confusion and complication. In the new changes, I've created new components for plotting visual progress tracks. Also, visual progress data no longer needs to be processed, and is in a consumable format right out of the box. The new shape of visual progress data looks like this `[{percent: 0, timestamp: 0}, {percent: 10, timestamp: 500}]`

2. I added more documentation on visual progress and visual metrics in the type files, including explanation on what visual metrics are, where does the data come from, and projects URLs.

Please let me know if I am missing anything. I look forward to your feedback.

gregtatum commented on src/components/timeline/GlobalTrack.js:
```diff
@@ -231,6 +280,7 @@ export default explicitConnect<OwnProps, StateProps, DispatchProps>({
     let localTrackOrder = EMPTY_TRACK_ORDER;
     let localTracks = EMPTY_LOCAL_TRACKS;
     let pid = null;
+    let graphData = [];
```
I think this should be `let graphData = null;` rather than empty data.

cliu55 replied:
This variable is initialized as null in the new update

gregtatum commented on src/components/timeline/TrackVisualProgressGraph.js:
```diff
@@ -0,0 +1,346 @@
+ * The visual progress track graph takes visual progress information from visual metrics, and renders it as a
+ * graph in the timeline.
+ */
+class TrackVisualProgressGraphImpl extends React.PureComponent<Props, State> {
```
I'm finding the combination of names `VisualProgress`, `VisualMetrics` and `graphData` to be a bit confusing. I generally find that having clear terminology helps the maintainability of the code.

It seems like `VisualMetrics` is all the data, and then there is a `VisualProgress` property, but the shape of the data is also called `VisualProgress`. The component is called `TrackVisualProgressGraph`. There is a `Prop` called `graphData`, but it's type is called `VisualProgress`.

Do you have any ideas here on how to make this more clear?

cliu55 replied:
Hi Greg, I'm thinking of renaming the shape and type of the data from `VisualProgress` to `ProgressGraphData`, and the `Prop` from `graphData` to `progressGraphData`. I think majority of the confusion comes from VisualProgress being a property and also a type.

gregtatum replied:
Thanks for thinking this through!

gregtatum commented on src/types/profile.js:
```diff
@@ -316,6 +316,42 @@ export type ExtensionTable = {|
+ * along with their visual progression. These properties find their way into the gecko profile through running
+ * browsertime, which is a tool that lets you collect performance metrics of your website.
+ * Browsertime provides the option of generating a gecko profile, which then embeds these visual metrics
+ * into the geckoprofile. More information about browsertime can be found through https://github.com/mozilla/browsertime.
```
This link is not providing me with information about what the individual values represent:

For instance: `PerceptualSpeedIndexProgress` isn't in the codebase.

I think it would be helpful to link to where this data gets generated. It's often nice to jump directly to the serializer to see what's going on.

cliu55 replied:
I've added links to where the data gets generated in the new update

gregtatum commented:
Oh, and codecov is a great place to see what your coverage looks like, the links are on this page.

<img width="693" alt="image" src="https://user-images.githubusercontent.com/1588648/62079372-33396400-b214-11e9-9c73-002a0ea3a29f.png">

In addition this can be tested locally by running `yarn test-coverage`

cliu55 commented on src/components/timeline/TrackVisualProgressGraph.js:
```diff
@@ -0,0 +1,346 @@
+|};
+
+type State = {|
+  hoveredCounter: null | number,
```
All the references to counter are now removed

cliu55 commented on src/components/timeline/TrackVisualProgressGraph.js:
```diff
@@ -0,0 +1,346 @@
+    }
+  };
+
+  _renderTooltip(counterIndex: number): React.Node {
```
All the references to counter are now removed

cliu55 commented on src/components/timeline/TrackVisualProgressGraph.js:
```diff
@@ -0,0 +1,346 @@
+   * Create a div that is a dot on top of the graph representing the current
+   * height of the graph.
+   */
+  _renderMemoryDot(counterIndex: number): React.Node {
```
All the references to memory are now removed

cliu55 commented on src/types/profile-derived.js:
```diff
@@ -157,7 +157,10 @@ export type StackType = 'js' | 'native' | 'unsymbolicated';
   | {| +type: 'process', +pid: Pid, +mainThreadIndex: ThreadIndex | null |}
-  | {| +type: 'screenshots', +id: string, +threadIndex: ThreadIndex |};
+  | {| +type: 'screenshots', +id: string, +threadIndex: ThreadIndex |}
+  | {| +type: 'visualProgress', +id: string |}
```
The new types have been changed to use kebab casing

cliu55 commented:
Hi @gregtatum,

Thank you for another review! In the newest updates, I've removed all references to memory and counter in variable naming in the new components, renamed some variable to avoid confusion ambiguity, and added new tests for the new component. Please let me know if you have further feedback on the changes.

gregtatum commented on src/types/profile.js:
```diff
@@ -357,15 +357,27 @@ export type VisualProgress = {|
-  PerceptualSpeedIndexProgress: VisualProgress[],
-  LastVisualChange: number,
+  // ContentfulSpeedIndex and ContentfulSpeedIndexProgress generated here
+  // https://github.com/mozilla/browsertime/blob/master/vendor/visualmetrics.py#L1331
```
Thanks for adding these. These aren't permalinks, and quickly grow stale. I would recommend going to the github page and hitting `Y` for a shortcut to get the permalink, or doing it through this menu:

<img width="358" alt="image" src="https://user-images.githubusercontent.com/1588648/64562771-465f4980-d313-11e9-8a25-ef3d9b7ece59.png">

gregtatum commented:
Oh, one other UI tweak:

<img width="300" alt="image" src="https://user-images.githubusercontent.com/1588648/64566331-5d099e80-d31b-11e9-94ec-fddbdaadeb41.png">

The text overflows here, it would be good to add the label as a `title` attribute as well, so that it displays the full text when hovering.
